### DISCUSSION TITLE
Failure on publishing project with HxGrid and HeaderTemplates

### DISCUSSION THREAD

**Discussion by michael-studenets** [havit/Havit.Blazor#514] [Q&A] [answered]
Have this error on publishing project. But when build and run it works.

The type or namespace name 'GridHeaderCellContext' does not exist in the namespace '__Blazor.Havit.Blazor.Components.Web.Bootstrap' (are you missing an assembly reference?)

**Comment by hakenr** [comment]
This type of error is usually caused by a syntax error in `.razor` file (and is not related to our component library).
For further assistance, paste your `Xy.razor` file here.

**Comment by pollardpj** [comment]
I get the same error, from a different angle:

Publishing in docker, using `FROM mcr.microsoft.com/dotnet/sdk:6.0-alpine AS build` as build image, with the Blazor project targeting net6.0, and Havit 3.2.10 (not tried with earlier versions, not sure it is relevant), I get the same error:

The type or namespace name 'GridHeaderCellContext' does not exist in the namespace '__Blazor.Havit.Blazor.Components.Web.Bootstrap' (are you missing an assembly reference?)

But if I change the build image to `FROM mcr.microsoft.com/dotnet/sdk:7.0-alpine AS build` I don't get this error, and the dotnet publish command succeeds.

I don't think there are any syntax issues in the razor.

I don't fully understand why this is the case, as I should be OK targetting 6.0 ... In the fullness of time it will become clear to me :) I did think it was something to do with the `record` key word, as GridHeaderCellContext is a record in Havit, but this may be a red herring.

This information might help, though - if I find anything more I'll add here.

**Comment by hakenr** [comment]
Any repro would be helpful to investigate further.

**Comment by pollardpj** [comment]
I did a repro for Blazor ASP.NET Hosted, publishing with docker: https://github.com/pollardpj/HxGrid_510

I think it may be possible to simplify this repro, for instance, I haven't tried on a machine without 7.0 sdk (Using docker in this way kind of achieves this).

Notes from this if you want to try it without cloning: 

Create a new Blazor Web Assembly project, ASP.NET Hosted, targetting net6.0
Add package Havit.Blazor.Components.Web.Bootstrap 3.2.10
Add @using Havit.Blazor.Components.Web.Bootstrap to _Imports.razor

Add class in Client project in Pages folder:

public class GridDataRow
{
    public string Value { get; set; }
}

Add to Program.cs in Client:

```
builder.Services.AddHxServices();
```

Add to Index.razor:

```
@page "/"

<PageTitle>Index</PageTitle>

<HxGrid TItem="GridDataRow" DataProvider="GetGridDataRowData" SelectionEnabled="false">
    <Columns>
        <HxGridColumn ItemTextSelector="@(context => context.Value)">
            <HeaderTemplate>
                <span>The Value</span>
            </HeaderTemplate>
        </HxGridColumn>
    </Columns>
</HxGrid>

@code {

    private async Task<GridDataProviderResult<GridDataRow>> GetGridDataRowData(GridDataProviderRequest<GridDataRow> request)
    {
        var results = new List<GridDataRow>
        {
            new GridDataRow { Value = "One" }
        };

        return await Task.FromResult(request.ApplyTo(results));
    }
}
```

Run the Blazor server project, and all works as expected.

Add docker support to the server project:

```
FROM mcr.microsoft.com/dotnet/aspnet:6.0 AS base
WORKDIR /app
EXPOSE 80
EXPOSE 443

FROM mcr.microsoft.com/dotnet/sdk:6.0 AS build
WORKDIR /src
COPY ["BlazorApp1/Server/BlazorApp1.Server.csproj", "BlazorApp1/Server/"]
COPY ["BlazorApp1/Client/BlazorApp1.Client.csproj", "BlazorApp1/Client/"]
COPY ["BlazorApp1/Shared/BlazorApp1.Shared.csproj", "BlazorApp1/Shared/"]
RUN dotnet restore "BlazorApp1/Server/BlazorApp1.Server.csproj"
COPY . .
WORKDIR "/src/BlazorApp1/Server"
RUN dotnet build "BlazorApp1.Server.csproj" -c Release -o /app/build

FROM build AS publish
RUN dotnet publish "BlazorApp1.Server.csproj" -c Release -o /app/publish /p:UseAppHost=false

FROM base AS final
WORKDIR /app
COPY --from=publish /app/publish .
ENTRYPOINT ["dotnet", "BlazorApp1.Server.dll"]
```

Try and build that docker image:

```
HxGrid_510\src> docker build -f .\BlazorApp1\Server\Dockerfile .
```

And it fails on;

```
error CS0234: The type or namespace name 'GridHeaderCellContext' does not exist in the namespace '__Blazor.Havit.Blazor.Components.Web.Bootstrap' (are you missing an assembly reference?)
```

If you change the build image to:

```
FROM mcr.microsoft.com/dotnet/sdk:7.0 AS build
```

It builds OK.

**Comment by pollardpj** [comment]
I have done some more digging, and I think it is related to https://github.com/dotnet/razor-compiler/pull/34 which is an issue to do with adding global:: to all namespaces in compiled razor. 

Best I can make out, with just .NET 6.0 SDK, you'll face this issue with some generic types (not specific to Havit). I think it's been sorted in 7.0.

@michael-studenets Perhaps you could cross-reference: On the machine/VM/whatever where you are publishing and seeing this error, what do you see when you type:

```
dotnet --list-sdks
```

**Comment by hakenr** [comment, marked as answer]
This is something we already met before and was a bug in .NET6 which was resolved in .NET7:
https://github.com/dotnet/aspnetcore/issues/37506 (reported by @jirikanda, my colleague from HAVIT team).

Unfortunately, I'm not ware of any workaround you could go over this issue other than using .NET7 (or not using the HeaderTemplate where the issue occurs) or defining the HeaderTemplate RenderFragment delegate from code, something like:
```razor
<HxGrid TItem="GridDataRow" DataProvider="GetGridDataRowData" SelectionEnabled="false">
	<Columns>
		<HxGridColumn ItemTextSelector="@(context => context.Value)" HeaderTemplate="RenderHeaderTemplate" />
	</Columns>
</HxGrid>

@code {
	private RenderFragment<GridHeaderCellContext> RenderHeaderTemplate => context => builder =>
	{
		builder.OpenElement(1, "span");
		builder.AddContent(2, "The Value");
		builder.CloseElement();
	};
}
```

(tested in this PR: https://github.com/pollardpj/HxGrid_510/pull/1)

**Comment by pollardpj** [comment]
Thank you for confirming where the issue lies @hakenr ! And thank you for the workaround, brilliant, helped me a lot and now understand a bit more :)
